filebeat/input/mqtt/input.go

```diff
@@ -0,0 +1,166 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package mqtt
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/filebeat/channel"
+	"github.com/elastic/beats/filebeat/input"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/backoff"
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/pkg/errors"
+
+	MQTT "github.com/eclipse/paho.mqtt.golang"
+)
+
+func init() {
+	err := input.Register("mqtt", NewInput)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Input contains the input and its config
+type mqttInput struct {
+	config        mqttInputConfig
+	context       input.Context
+	outlet        channel.Outleter
+	log           *logp.Logger
+	mqttWaitGroup sync.WaitGroup
+	runOnce       sync.Once
+	client        MQTT.Client
+}
+
+// NewInput creates a new mqtt input
+func NewInput(
+	cfg *common.Config,
+	connector channel.Connector,
+	inputContext input.Context,
+) (input.Input, error) {
+
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrap(err, "reading mqtt input config")
+	}
+
+	out, err := connector.ConnectWith(cfg, beat.ClientConfig{
+		Processing: beat.ProcessingConfig{
+			DynamicFields: inputContext.DynamicFields,
+		},
+		//		ACKEvents: func(events []interface{}) {
+		//			for _, event := range events {
+		//				if meta, ok := event.(eventMeta); ok {
+		//					meta.handler.ack(meta.message)
+		//				}
+		//			}
+		//		},
+		CloseRef:  doneChannelContext(inputContext.Done),
+		WaitClose: config.WaitClose,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	input := &mqttInput{
+		config:  config,
+		context: inputContext,
+		outlet:  out,
+		log:     logp.NewLogger("mqtt input").With("host", config.Host),
+	}
+
+	input.setupMqttClient()
+
+	return input, nil
+}
+
+// Run starts the input by scanning for incoming messages and errors.
+func (input *mqttInput) Run() {
+	input.runOnce.Do(func() {
+		go func() {
+
+			context := doneChannelContext(input.context.Done)
```

Comment:
`doneChannelContext` is probably not the best choice here. That wrapper exists because a library dependency expected a context to manage timeouts / cancellation, but here that isn't the case so we can fall back on the simpler pattern of using the `Done` channel directly.

Reply:
I removed the context now. Let me know if you have a more robust idea in mind for the connection procedure.